kyo-core/shared/src/main/scala/kyo/chunks.scala

```diff
@@ -0,0 +1,464 @@
+            case c: Compact[T] =>
+                c.array
+            case c =>
+                c.toArray(using ClassTag(classOf[Any]))
+
+    override def equals(other: Any) =
+        (this eq other.asInstanceOf[Object]) || {
+            other match
+                case other: Chunk[?] =>
+                    (this.isEmpty && other.isEmpty) ||
+                    (this.size == other.size && {
+                        val s = this.size
+                        val a = this.toIndexed
+                        val b = other.toIndexed
+                        @tailrec def loop(idx: Int = 0): Boolean =
+                            if idx < s then
+                                inline given CanEqual[Any, Any] = CanEqual.derived
+                                a(idx) == b(idx) && loop(idx + 1)
+                            else
+                                true
+                        loop()
+                    })
+                case _ =>
+                    false
+        }
+
+    override def hashCode(): Int =
+        if isEmpty then
+            1
+        else
+            val s = this.size
+            val a = this.toIndexed
+            @tailrec def loop(idx: Int = 0, acc: Int = 1): Int =
+                if idx < s then
+                    loop(idx + 1, 31 * acc + a(idx).hashCode())
+                else
+                    acc
+            loop()
+end Chunk
+
+object Chunks:
+
+    import internal.*
+
+    sealed abstract class Indexed[T] extends Chunk[T]:
+
+        //////////////////
+        // O(1) methods //
+        //////////////////
+
+        def apply(i: Int): T
+
+        final def head: T =
+            if isEmpty then
+                throw new NoSuchElementException
+            else
+                apply(0)
+
+        final def tail: Indexed[T] =
+            if size <= 1 then cachedEmpty.asInstanceOf[Indexed[T]]
+            else
+                this match
+                    case Tail(chunk, offset, size) =>
+                        Tail(chunk, offset + 1, size - 1)
+                    case c =>
+                        Tail(c, 1, size - 1)
+    end Indexed
+
+    def init[T]: Chunk[T] =
+        cachedEmpty.asInstanceOf[Chunk[T]]
+
+    def init[T](values: T*): Chunk[T] =
+        initSeq(values)
+
+    def initSeq[T](values: Seq[T]): Chunk[T] =
+        if values.isEmpty then init[T]
+        else
+            values match
+                case seq: IndexedSeq[T] => FromSeq(seq)
+                case _                  => Compact(values.toArray(using ClassTag(classOf[Any])))
+
+    def fill[T, S](n: Int)(v: T): Chunk[T] =
+        if n <= 0 then Chunks.init
+        else
+            val array = (new Array[Any](n)).asInstanceOf[Array[T]]
+            @tailrec def loop(idx: Int = 0): Unit =
+                if idx < n then
+                    array(idx) = v
+                    loop(idx + 1)
+            loop()
+            Compact(array)
+    end fill
+
```

Comment:
Perhaps worth adding a `tabulate` method? ```def tabulate[T](n: Int)(f: Int => T): Chunk[T] =```

Reply:
I think there are many methods we'll need to add yet but I'd prefer to keep the scope of this PR to the current methods. I need this change to open the PR with the new chunked streams.